docs/ql-style-guide.md

```diff
@@ -0,0 +1,434 @@
+
+  override string toString() { result = "Throw Exception" }
+}
+```
+
+## Spaces
+1. There *must* be a space or line break:
+   - surrounding each `=` and `|`,
+   - after each comma.
+1. There *should* be a space or line break:
+   - surrounding each *binary operator*, which *must* be balanced,
+   - surrounding `..` in a range.
+   - Exceptions to this are to save space or to improve readability.
+1. *Avoid* other spaces, for example:
+   - after a *quantifier/aggregation* keyword,
+   - after the predicate name in a *call*,
+   - inside brackets used for *calls*, single-line quantifiers, and parenthesised formulas,
+   - surrounding a `.`,
+   - inside the opening or closing `[ ]` In a range expression,
+   - inside casts `a.(X)`.
+1. *Avoid* multiple spaces, except for indentation, and *avoid* additional indentation to align formulas, parameters or arguments.
+1. *Do not* put whitespace on blank lines, or trailing on the end of a line.
+1. *Do not* use tabs.
+
+
+For example,
+
+```ql
+cached
+private predicate foo(Expr e, Expr p) {
+  exists(int n |
+    n in [0 .. 1] |
+    e = p.getChild(n + 1)
+  )
+}
+```
+
+## Naming
+1. Use PascalCase for:
+   - `class` names,
+   - `module` names,
+   - `newtype` names.
+1. Use camelCase for:
+   - predicate names,
+   - variable names.
+1. Newtype predicate names *should* begin with `T`.
+1. *Avoid* underscores in names.
+1. Predicates that have a result *should* be named `get...`
+1. Predicates that can return multiple results *should* be named `getA...` or `getAn...`
+1. Predicates that don't have a result or parameters *should* be named `is...` or `has...`
+1. *Avoid* short or single-letter names for classes, predicates and fields.
+1. Short or single letter names for parameters and *quantifiers* *may* be used provided that they are sufficiently clear.
+1. Use names as they are used in the target-language specification.
+
+Examples:
+
+```ql
+/** ... */
+predicate calls(Callable caller, Callable callee) {
+  ...
+}
+```
+
+```ql
+/** ... */
+class Type extends ... {
+  /** ... */
+  string getName() { ... }
+
+  /** ... */
+  predicate declares(Member m) { ... }
+
+  /** ... */
+  predicate isGeneric() { ... }
+
+  /** ... */
+  Type getTypeParameter(int n) { ... }
+
+  /** ... */
+  Type getATypeParameter() { ... }
+}
+```
+
+## Documentation
+1. Documentation *must* adhere to the [QLDoc specification](https://help.semmle.com/QL/QLDocSpecification.html).
+1. Public declarations *must* be documented.
+1. Non-public declarations *should* be documented.
+1. Declarations in query files *should* be documented.
+1. Use `/** ... */` for documentation, even for single line comments.
+1. For single-line documentation, the `/**` and `*/` are written on the same line as the comment.
+1. For multi-line documentation, the `/**` and `*/` are written on separate lines. There is a `*` preceding each comment line, aligned on the first `*`.
+1. Library files (`.qll` files) *should* be documented with QLDoc.

```

Comment:
What does this mean?

Reply:
It means there's a documentation comment at the top of the file. Will clarify.